pr47-core/src/vm/al31f/executor/async_executor.rs

```diff
@@ -148,45 +152,101 @@ unsafe fn checked_exception_unwind_stack<A: Alloc>(
     Err(exception)
 }
 
-async unsafe fn vm_thread_run_function_impl<A: Alloc>(
-    thread: &mut VMThread<A>,
-    func_id: usize,
-    args: &[Value]
-) -> Result<Vec<Value>, Exception> {
-    #[cfg(feature = "bench")] let start_time: std::time::Instant = std::time::Instant::now();
-    #[cfg(feature = "bench")] defer!(move || {
-        let end_time: std::time::Instant = std::time::Instant::now();
-        eprintln!("Time consumed: {}ms", (end_time - start_time).as_millis());
-    });
+pub struct VMThreadRunFunctionFut<'a, A: Alloc, const S: bool> {
+    thread: &'a mut VMThread<A>,
+    slice: StackSlice,
+    insc_ptr: usize,
 
-    get_vm!(thread).alloc.set_gc_allowed(true);
+    awaiting_promise: Option<Pin<Box<dyn Future<Output = AsyncReturnType>>>>
```

Comment:
Here comes the question: could we avoid using such a `Pin<Box>`?

Reply:
I guess that futures about `Mutex`es should be mostly `Unpin`. The only problem is how could we accomplish type erasure.

Reply:
There's now one thing for sure: we could neither explicitly write out the type, nor could we erase it without allocation.